### DISCUSSION TITLE
Cross program calls: examples

### DISCUSSION THREAD

**Discussion by matthiasgoergens** [0xmozak/mozak-vm#1324] [General]
Here I am collecting some examples of patterns of call graphs we should support.

They are not necessarily meant to be practical, but prototypical of patterns that might occur in practical programs.

# Parity checker

Assume we have two cooperating programs, that call each other via cross program calls:

```rust
fn check_even(x: u32) -> bool {
  x == 0 || check_odd(x - 1)
}
```

```rust
fn check_odd(x: u32) -> bool {
  x != 0 && check_even(x - 1)
}
```

Crucially, `check_even` (and `check_odd`) typically indirectly calls themselves before returning anything.

`check_odd(4)` would result in the following calls on the tape (in some order):
```rust
native     =>  check_odd(4)  = false
check_odd  => check_even(3) = false
check_even =>  check_odd(2) = false
check_odd  => check_even(1) = false
check_even =>  check_odd(0) = false
```

# Fibonacci

The obligatory Fibonacci example.

```rust
fn fib(x: u32) -> u32 {
  if x < 2 {
    x
  } else {
    fib(x - 1) + fib(x - 2)
  }
}
```

Assume that calls to `fib` happen via cross-program calls.  (In a real program, there might be some intervening other programs involved, instead of direct self calls.)

**Comment by supragya** [comment]
# How things work for Parity Checker
Assume we invoke for `x = 3` from a top level caller that calls `check_odd`.
Under native execution, this would generate a call tape as follows:
```
1 [from: NULL   TO: OddPrg    FN: CheckOdd  Arg: 3   Ret: True]
2 [from: OddPrg TO: EvnPrg    FN: CheckEvn  Arg: 2   Ret: True]
3 [from: EvnPrg TO: OddPrg    FN: CheckOdd  Arg: 1   Ret: True]
4 [from: OddPrg TO: EvnPrg    FN: CheckEvn  Arg: 0   Ret: True]
```

Programs `OddPrg` and `EvnPrg` are run separately.
When `OddPrg` runs, it walks through each calls that has `TO:` as `OddPrg`. That in this case would be `1`, `3`.

When running `1`, `OddPrg` executed `CheckOdd`, which internally calls `CheckEvn` with certain arguments. Let's say these runtime args is `a`. Since `1` is where the call to `CheckOdd` was, the next in line, `1+1 = 2` is where `CheckEvn` call should feature. `CheckOdd` checks if at index `2`, a call to same program `EvnPrg` and same function `CheckEvn` and same generated argument `a` is featured. If yes, `CheckOdd` happily takes the ret value mentioned in the index `2`, executes
```rust
x != 0 && check_even(x - 1)
// resolves to true && true
```
ensures ret at `1` matches this value and completes it's validation checks for `1`.

Side effect of above is that while checking for index `1`, we got to know that index `2` is subpart of the execution under `1`. Hence, a loop that walks over the CallTape to find calls that has `TO:` as `OddPrg` will skip `2` and directly check `3` next where similar resolution occurs.

At the end, `OddPrg` ensures all calls to itself has been responded by it in CallTape correctly.


Same kind of execution would be done by `EvnPrg`, hence both programs validate the CallTape.

**Comment by supragya** [comment]
# Protocol extension for self-calls under Cross-Program-Calls
All of the current design of cross-program-calls have an unmentioned axiom inherent in the design. It could be described as follows:

> cross-program-call is actually to be only used when calls are supposed to exist between two distinct program, i.e. the caller and the callee programs are supposed to be different.

However, in a situation where the following is expected:
> Extend CPC design to remove the restriction between the caller and the callee such that CPC can be opted in even if internal dispatch would have sufficed.

an extension for self-calls can be introduced in CPC.

The general idea remains the same as before, but extended enough to accomodate for self-calls. We describe the extension as a simple operation where each caller and callee instead of being identified by their `ProgramIdentifier`, they instead are now identified by `(ProgramIdentifier, self_call_depth)`.

In parity example above where self-call wasn't introduced, if extension is applied, the CallTape would simply change from:
```
1 [from: NULL   TO: OddPrg    FN: CheckOdd  Arg: 3   Ret: True]
2 [from: OddPrg TO: EvnPrg    FN: CheckEvn  Arg: 2   Ret: True]
3 [from: EvnPrg TO: OddPrg    FN: CheckOdd  Arg: 1   Ret: True]
4 [from: OddPrg TO: EvnPrg    FN: CheckEvn  Arg: 0   Ret: True]
```
to:
```
1 [from: (NULL  , 0) TO: (OddPrg, 0)    FN: CheckOdd  Arg: 3   Ret: True]
2 [from: (OddPrg, 0) TO: (EvnPrg, 0)    FN: CheckEvn  Arg: 2   Ret: True]
3 [from: (EvnPrg, 0) TO: (OddPrg, 0)    FN: CheckOdd  Arg: 1   Ret: True]
4 [from: (OddPrg, 0) TO: (EvnPrg, 0)    FN: CheckEvn  Arg: 0   Ret: True]
```

and instead of validation logic being:
```rust
for &message in calltape.iter() {
    if message.callee == self_prog_id {
        // Process
    }
}
```
the check becomes
```rust
for call_depth in 0..calltape.max_self_call_depth(self_prog_id) {
    for &message in calltape.iter() {
        if message.callee == self_prog_id {
            // Process
        }
    }
}
```
The assumption is that `calltape.max_self_call_depth(self_prog_id)` takes no extra time to calculate since a single traversal of CallTape is anyways needed to setup the `Vec<ArchivedCPCMessage>` in zkvm.

The extension makes us of `self_call_depth` as follows:

"Whenever a program at any specific time does a CPC call to itself, (self-call), the call is registered in the CallTape with caller: `(Program, id)` and callee: `(Program, id+1)` where `id` is a number signifying call depth. If however, the programs are distinct, the caller remains `(Program, id)` but callee becomes `(OtherProgram, 0)` independent of the `id`."

This remains the general idea of the self-call extension. Further, the fibonacci solution would present more concrete example on how call-tape looks, and the additional lookup load on accessing CPC tape with such extension.

**Reply by matthiasgoergens** [reply]
This seems like a weird ad hoc design.  Let me think some more, but it 'smells' funny.

**Comment by supragya** [comment]
# # How things work for Fibonacci

We extend the above given fibonacci example to bring home the implications of "self-calls" extension to CPC. For this, we make of two cooperating programs defined as follows:

```rust
// PROGRAM: AlfProg
fn alf(x: u32) -> u32 {
    if x > 0 {
        fib(x)
    }
    0
}
```
and
```rust
// PROGRAM: FibProg
fn fib(x: u32) -> u32 {
    if x < 2 {
        x + alf(0)
    } else {
        fib(x - 1) + fib(x - 2)
    }
}
```

Assume, we initiate the call from `NULL` as `alf(4)`. If you generate the call-tree (if you may), it would look as follows (assume presentation as `-> fun(arg) call_depth` e.g. `-> fib(4) 1` is call to `fib` with arg `4` at `call_depth == 1`):
```
-> alf(4) 0
-> fib(4) 0                                                     
-> fib(3) 1                      -> fib(2)  1                   
-> fib(2) 2    -> fib(1) 2       -> alf(0)  0           
-> alf(0) 0    -> alf(0) 0
```
The calltape now looks like (space added for readability):

```
1. [from: NULL          TO: (AlfProg, 0)    FN: alf  Arg: 4   Ret: 5     next: 2]

2. [from: (AlfProg, 0)  TO: (FibProg, 0)    FN: fib  Arg: 4   Ret: 5     next: 3]

3. [from: (FibProg, 0)  To: (FibProg, 1)    FN: fib  Arg: 3   Ret: 3     next: 5]  ------|
4. [from: (FibProg, 0)  To: (FibProg, 1)    FN: fib  Arg: 2   Ret: 2     next: 7]  ---|  |
                                                                                      |  |
5. [from: (FibProg, 1)  To: (FibProg, 2)    FN: fib  Arg: 2   Ret: 2     next: 8]  <------
6. [from: (FibProg, 1)  To: (FibProg, 2)    FN: fib  Arg: 1   Ret: 1     next: 9]     | 
                                                                                      |
7. [from: (FibProg, 1)  To: (AlfProg, 0)    FN: alf  Arg: 0   Ret: 0     next: -]    <-

8. [from: (FibProg, 2)  To: (AlfProg, 0)    FN: alf  Arg: 0   Ret: 0     next: -]
9. [from: (FibProg, 2)  To: (AlfProg, 0)    FN: alf  Arg: 0   Ret: 0     next: -]
```

We know that while executing `AlfProg` at calltape index `0`, during execution, `AlfProg` will know that internally it would do just `1` call, i.e. idx `2`. So, after checking `1`, the next check for `callee` only happens at idx `3`. Let's call it the "skip-trick" for now.
Also, `AlfProg` knows that his `calltape.max_self_call_depth(self_prog_id)` value is `1`.

So, for `AlfProg` under following code
```rust
for call_depth in 0..calltape.max_self_call_depth(self_prog_id) {  // 1
    for &message in calltape.iter() {
        if message.callee == self_prog_id {
            // Process  
        }
    }
}
```
calltape is only traversed once.

### But what is `next` ?
Look closely, the CallTape is just a level-order traversal of the call-tree. `next` in essence just encodes where to jump next if you want to "resolve". You do not use `next` when you are not "resolving".

For example, look at `3`. When "resolving" `3`, the value `next` mentioned that one needs to go to `5`. So any internal calls would be `5+0`, `5+1`, ..., `5+N` in sequence if there are `N` internal calls.
However, the "skip-trick" is now a bit more complicated (caveat: you can still skip `5+0` through `5+N`, but they are just not adjecent to you).

Where does this come from? This can pretty much be injected by native as a hint for resolution inside zkvm. Any wrong values will anyways break the validation, and there is no loss of security here by adding wrong hints.

### The downside
Doing the CPC call to oneself quite evidently will now require multiple traversals on the call-graph. `calltape.max_self_call_depth(self_prog_id)` to be exact. Without optimizations, every self-call would make traversal for the program doing self-calls become `O(D*N)` where `D` is `calltape.max_self_call_depth(self_prog_id)` and `N` is the size of the CallTape.

However, do note that the additional penalty is only for programs doing self-calls and not any other program involved. `AlfProg` in above example still only traverses the CallTape once.

### Conclusion
While self-call extension is cute, the additional penalty in CPC traversal may not be worth it. One has to contain max-depth. Still, this can be provided as a feature within SDK.

**Reply by supragya** [reply]
In essence, prefer internal dispatch over CPC

**Reply by supragya** [reply]
Interesting addition by @matthiasgoergens: `We don't need the call depth to enable self calls. Just two Indices.` Yes this reduces it to max `O(2*N)` much more palateable

**Reply by matthiasgoergens** [reply]
@supragya What happened to this design?  I thought that was what we agreed on, not this call depth counter (nor a `mod 2` version of it.)
